### ISSUE TITLE
composer.json in dev branch

### ISSUE THREAD

**Issue by WinterSilence** [koseven/koseven#360]
~~~
  "require": {
    "php": ">=7.3",
    "ext-dom": "*"
  },
  "require-dev": {
    "phpunit/phpunit": "^8",
    "php-coveralls/php-coveralls": "^2.0",
    "ext-curl": "*"
},
~~~

1. Please dont change min version for PHP/libraries without testing and update code (current module `unittest` not support phpunit 8)
2.  `8.*` better `^8` - we cant check vendor releases every day
3. Support legacy versions is bad practice, but set latest versions as minimal its bad too.

I vote set `"php": "7.*"`, `"phpunit/phpunit": "7.*"`

**Comment by toitzi** [comment]
@WinterSilence Thanks for Feedback. I am very sorry for my long response but i will take this as oppertunity to discuss some/a big thing(s).

> Please dont change min version for PHP/libraries without testing and update code (current module unittest not support phpunit 8)

Tested the code ofc and also travis ci build passed. Module unittest in it's updated Version (devel) should work, if not please post those tests as i will of course fix them. Thanks for noticing!

> 8.* better ^8 - we cant check vendor releases every day

Agree, ill take a look.

> Support legacy versions is bad practice, but set latest versions as minimal its bad too.

Correct but since we were on PHPUnit 6 and the jump from 6 to 8 was not that much to change i thought it would be the best time to do the upgrade instead of beeing 1 Version behind again. Also when it comes to Unittesting i rather stick to newer versions than legacy, because sticking to legacy phpunit caused us more troubles and work then it takes to upgrade to 8. So definitley a +1 for Version 8 from me.

> I vote set "php": "7.*", "phpunit/phpunit": "7.*"

My opinion why we should stick to version 8 of PHPUnit is explained above. And as for PHP:
Our next release Version is **Koseven 4** which i am very excited about to come as soon as possible. (maybe even prior our roadmap). Anyways if you look at the [Roadmap](https://github.com/koseven/koseven/blob/master/ROADMAP.md) you'll see that the "targeted" PHP-Version is 7.3+ and in fact that's why i changed it to 7.3+

Both of above has to happen at some point so in fact what i get from this is that we have to decide how big the step to 4.0 will be. I mean as our Roadmap says (i mean it is common practice) with all major releases an API break occurs. As said remaining questions: How much shall we break / upgrade, what is too far what is too far behind.

Possible Solutions / Votes i can see:

1. Downgrade back to PHPUnit 7+, Make not use of features introduced in PHP 7.3+ to support older 
    PHP Versions - _In favor of compatibility_
2. Keep PHPUnit 8+ (and fix the `unittest` module), Take advantage from the Major release and make 
    Koseven use of newer PHP Version(s) - _In favor of features, speed, etc_
3. Keep PHPUnit 8+ (and fix the `unittest` module),  Make not use of features introduced in PHP 7.3+ 
    to support older PHP Versions (PHPUnit only supports till 7.3) - _Compatibility 
   from supporting older PHP Versions and keeping our Unittests up to date_

Personally i am voting for number **2** (Although i could live with all those 3 possibilities just fine). I can very much see the reason of your concern according to break more backwards compatibility, but i will give you **my** opinion why we should take the chance to upgrade:

Version 4.0 will be the **first** major release Versions since forking Kohana (which was already deprecated and **very** behind in terms of technology in 2017). Since than Koseven did not make really major changes to that and was "stuck" on the old technology Kohana used (which is absolutly fine because it was aimed to be "replacement for Kohana" without really changing anything / much).
Today we got 2019 and technology that was behind in 2017 is even far more behind today. 
And instead of beeing behind and put work into "upgrading" to legacy / soon legacy components again would be much more work to maintain and in my eyes.....simply not worth it. I think we should make use of the new major release to bring Koseven forward to today technology (no worries i understood that you did not want it to run on php5+ again but anyway). Instead of maintaining old libraries which are outdated the time we start using it, i think we should move one step forward. Time maintaining old libraries, we could invest in things like "basic" features that lack in koseven right now, unittests, documentation, etc... I am really excited to discuss this a bit since it would also give me a clearer perspective of what koseven shall be in the future, but as said in my Opinion it's time to take the cance and do a big step.

Excited to hear your opinion on that @neo22s @WinterSilence @jstrobel @svenbw @piotrbaczek .

**Comment by toitzi** [comment]
Update: And as i forgot: Keep in mind that Koseven has been compatible to Kohana code and provided support since almost 2 years. I think that is enough time for everyone to switch framework or keep koseven and update their code accordingly to it. But if we don't "break free" now from that we may stick to technology which is behind forever.

**Comment by WinterSilence** [comment]
@toitzi i vote for 1
7.2-7.3 dont adds anything important. For new functions we may add temporary solutions like as `if ( ! function_exists('...')) {`
We can go step by step to new versions of PHP: update version only if really need for new version KO7
https://w3techs.com/technologies/details/pl-php/7/all - as u can see only 5% can use new KO7, at release maybe 10-15%
Sad but true: my commerce solutions `>=5.6` because 70% still use php5

**Comment by WinterSilence** [comment]
@toitzi 
> Today we got 2019 and technology that was behind in 2017 is even far more behind today.

https://github.com/koseven/koseven/issues/348#issuecomment-487473049

**Comment by toitzi** [comment]
> i vote for 1

Specific reason you also want to downgrade PHPUnit? If you look at the [changelog](https://phpunit.de/announcements/phpunit-8.html), you will notice that the only breaking change here is the return type for 7 functions to `void`....which can easily be adapted in existing code. I think it's not too much to adapt to that and use it in favor of newer version.

> 7.2-7.3 dont adds anything important.

Even those are the ones used in this issue and as example, my discussion was more generalized (not only php also other components) and just wanted to get clear about that in the future / for future releases.

> #348 (comment)

Not really what i meant. I don't have time to maintain second fork with different strategies, etc...in your comment you talk about "kohana" recognition, as i said code has to be updated at some point and i think 2 years is enough time which we kept it "old". Updating to newer technology does not make Koseven/Kohana "less recognize-able" since we stick to the same principles (CSF, Conventions, Routing System, etc...)

**Comment by WinterSilence** [comment]
> Tested the code ofc and also travis ci build passed. 

Versions not fixed, example: u push PR at phpunit 8.1, after this phpunit updated to 8.2 and I install other pack

> Module unittest in it's updated Version (devel) should work, if not please post those tests as i will of course fix them.

My mistake: dev server use locale `ru_RU`. 

> Specific reason you also want to downgrade PHPUnit?

> https://w3techs.com/technologies/details/pl-php/7/all - as u can see only 5% can use new KO7, at release maybe 10-15% (as example my host with max version 7.2)

Who can use KO7 with PHP 7.3?  5-10%? Framework without users.. "KO7 too modern for you, come later" :)  phpunit 7 work at PHP `>=7.0` and covers more users/hosts. framework fornot se

**Comment by toitzi** [comment]
@WinterSilence 

> Versions not fixed, example: u push PR at phpunit 8.1, after this phpunit updated to 8.2 and I install other pack

Fixed them in my current PR will be updated soon, thanks for recognizing.

> Who can use KO7 with PHP 7.3? 5-10%? Framework without users.. "KO7 too modern for you, come later" :) phpunit 7 work at PHP >=7.0 and covers more users/hosts

I think "can" is the wrong wording here. Marketshare of 7.0 or 7.3 is not depending on "can use" it is depending on software which runs on it (That's why the "market share" of 7.3 incrases from time to time). If you are self-hosting there is no issue of running PHP 7.3 instead 7.0 (Only if you run an old OS). If you are not self-hosting the most control panels and hosting provides support 7.3
 nowdays.
In both cases you can even run parallel php versions if you need software/apps that run on older versions.

> Legacy code with new technology look awful.

Agree that's why we try to update it asap.

**Anyway**:
After discussing this with other members i think we will use what we planned to do in first place and stick with PHP7.3 - like our [Roadmap](https://github.com/koseven/koseven/blob/master/ROADMAP.md) says.

Feel free to re-open and share your thoughts if neccessary.
